### ISSUE TITLE
[GQSP] Computing complementary polynomial by root-finding is very inaccurate

### ISSUE THREAD

**Issue by anurudhp** [quantumlib/Qualtran#631]
Theorem 4 of https://arxiv.org/abs/2308.01501: Computes roots of an auxiliary polynomial R (from P) which are then re-interpolated to construct Q. 

https://github.com/quantumlib/Qualtran/blob/642dec21cb6844bbe2f5f33e4312a4951f73e8c9/qualtran/bloqs/generalized_qsp.py#L82-L187

Interestingly, the computed Q satisfies the constraint $|P(z)|^2 + |Q(z)|^2 = 1$ for polynomials upto degree 100. But the QSP sequence is only accurate till degree 4 (related to #630).

**Comment by tanujkhattar** [comment]
Can you add a small example of a `P` for which you compute the corresponding valid `Q` but the QSP sequence turns out to be incorrect due to precision issue in computing phase angles?

**Comment by anurudhp** [comment]
```py
P = [0.07691151711056517, 0.02506952291383959, 0.02858956904068647, 0.08523950878413064, 0.09750064936065875, 0.08848532934911055, 0.035950784393690226, 0.05988589458757472, 0.03547956116572998, 0.03401902153706458]
Q = [-0.002674721864062315, -0.00375596522126604, -0.006826640697387547, -0.008793329571296443, -0.01680094202314425, -0.024007293148226038, -0.027211524947234132, -0.029928727434094398, -0.03458906260723875, 0.9782155640881396]
```

code to reproduce test:
```py
def test_custom():
    random_state = np.random.RandomState(100)

    U = RandomGate.create(2, random_state=random_state)
    P = random_qsp_polynomial(degree=10, random_state=random_state, only_real_coeffs=True)
    Q = GeneralizedQSP(U, P).Q
    check_polynomial_pair_on_random_points_on_unit_circle(
        P, Q, random_state=random_state, n_points=10**5
    ) # works

    verify_generalized_qsp(U, P) # fails, error is ~ 0.00166
```

**Comment by tanujkhattar** [comment]
For the example you posted, the problem doesn't seem to be related to precision issues in computing phase angles as highlighted in https://github.com/quantumlib/Qualtran/issues/630

For example, when computing the phase angles for the above example, the values of a & b are as follows:

```
d=4, phi[d]= 0.0 a= (0.1+0j) b= (0.8996+0j) a/b= (0.1112+0j)
d=3, phi[d]= 0.0 a= (0.9051+0j) b= (0.3128+0j) a/b= (2.8935+0j)
d=2, phi[d]= 3.1416 a= (0.9577+0j) b= (-0.1365-0j) a/b= (-7.0171+0j)
d=1, phi[d]= 3.1416 a= (-0.9288+0j) b= (0.2119+0j) a/b= (-4.3826+0j)
d=0, phi[d]= 3.1416 a= (0.8584-0j) b= (-0.0811+0j) a/b= (-10.5883+0j)
```

None of the a, b or a/b seems to be close to 0. Do you have a better understanding of what may be causing the error in final unitary? Also, do we have examples of P & Q (not necessarily generated using `qsp_complementary_polynomial`) for which `qsp_phase_factors` (algorithm 1) implementation works as expected for higher degree polynomials?

**Comment by anurudhp** [comment]
fixed by #636
